Private opaque type in class

## Compiler version

3.0.0...3.1.2-RC2

## Minimized code

```scala
class Bar:
  private opaque type Baz = Int
```

## Output (click arrow to expand)
<details>

```scala
error while transforming type Baz = Int
error while transforming () extends Object() {
  type Baz = Int
}
error while transforming @SourceFile("/var/folders/rp/f9y80jvs54lcy26rxnt2y2kw0000gn/T/NoOwner.scala") 
  opaque
 class Bar() extends Object() {
  type Baz = Int
}
error while transforming package <empty> {
  @SourceFile("/var/folders/rp/f9y80jvs54lcy26rxnt2y2kw0000gn/T/NoOwner.scala") 
    opaque
   class Bar() extends Object() {
    type Baz = Int
  }
}
exception occurred while compiling /var/folders/rp/f9y80jvs54lcy26rxnt2y2kw0000gn/T/NoOwner.scala
java.lang.AssertionError: NoDenotation.owner while compiling /var/folders/rp/f9y80jvs54lcy26rxnt2y2kw0000gn/T/NoOwner.scala
Exception in thread "main" java.lang.AssertionError: NoDenotation.owner
        at dotty.tools.dotc.core.SymDenotations$NoDenotation$.owner(SymDenotations.scala:2495)
        at dotty.tools.dotc.core.Symbols$Symbol.retainsDefTree(Symbols.scala:87)
        at dotty.tools.dotc.core.Symbols$Symbol.defTree_$eq(Symbols.scala:79)
        at dotty.tools.dotc.transform.PostTyper$PostTyperTransformer.processMemberDef(PostTyper.scala:142)
        at dotty.tools.dotc.transform.PostTyper$PostTyperTransformer.transform(PostTyper.scala:386)
        at dotty.tools.dotc.transform.MacroTransform$Transformer.transformStat$1(MacroTransform.scala:39)
        at dotty.tools.dotc.transform.MacroTransform$Transformer.transformStats$$anonfun$1(MacroTransform.scala:42)
        at dotty.tools.dotc.core.Decorators$ListDecorator$.loop$1(Decorators.scala:92)
        at dotty.tools.dotc.core.Decorators$ListDecorator$.mapconserve$extension(Decorators.scala:108)
        at dotty.tools.dotc.transform.MacroTransform$Transformer.transformStats(MacroTransform.scala:42)
        at dotty.tools.dotc.transform.PostTyper$PostTyperTransformer.transformStats(PostTyper.scala:460)
        at dotty.tools.dotc.transform.MacroTransform$Transformer.transform(MacroTransform.scala:58)
        at dotty.tools.dotc.transform.PostTyper$PostTyperTransformer.transform$$anonfun$4$$anonfun$1(PostTyper.scala:351)
        at dotty.tools.dotc.transform.SuperAccessors.wrapTemplate(SuperAccessors.scala:210)
        at dotty.tools.dotc.transform.PostTyper$PostTyperTransformer.transform$$anonfun$4(PostTyper.scala:351)
        at dotty.tools.dotc.transform.PostTyper$PostTyperTransformer.withNoCheckNews(PostTyper.scala:98)
        at dotty.tools.dotc.transform.PostTyper$PostTyperTransformer.transform(PostTyper.scala:353)
        at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform(Trees.scala:1466)
        at dotty.tools.dotc.transform.MacroTransform$Transformer.transform(MacroTransform.scala:51)
        at dotty.tools.dotc.transform.PostTyper$PostTyperTransformer.transform(PostTyper.scala:386)
        at dotty.tools.dotc.transform.MacroTransform$Transformer.transformStat$1(MacroTransform.scala:39)
        at dotty.tools.dotc.transform.MacroTransform$Transformer.transformStats$$anonfun$1(MacroTransform.scala:42)
        at dotty.tools.dotc.core.Decorators$ListDecorator$.loop$1(Decorators.scala:92)
        at dotty.tools.dotc.core.Decorators$ListDecorator$.mapconserve$extension(Decorators.scala:108)
        at dotty.tools.dotc.transform.MacroTransform$Transformer.transformStats(MacroTransform.scala:42)
        at dotty.tools.dotc.transform.PostTyper$PostTyperTransformer.transformStats(PostTyper.scala:460)
        at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform(Trees.scala:1477)
        at dotty.tools.dotc.transform.MacroTransform$Transformer.transform(MacroTransform.scala:51)
        at dotty.tools.dotc.transform.PostTyper$PostTyperTransformer.transform(PostTyper.scala:451)
        at dotty.tools.dotc.transform.MacroTransform.run(MacroTransform.scala:21)
        at dotty.tools.dotc.core.Phases$Phase.runOn$$anonfun$1(Phases.scala:308)
        at scala.collection.immutable.List.map(List.scala:246)
        at dotty.tools.dotc.core.Phases$Phase.runOn(Phases.scala:309)
        at dotty.tools.dotc.Run.runPhases$1$$anonfun$1(Run.scala:259)
        at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
        at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
        at scala.collection.ArrayOps$.foreach$extension(ArrayOps.scala:1328)
        at dotty.tools.dotc.Run.runPhases$1(Run.scala:270)
        at dotty.tools.dotc.Run.compileUnits$$anonfun$1(Run.scala:278)
        at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
        at dotty.tools.dotc.util.Stats$.maybeMonitored(Stats.scala:68)
        at dotty.tools.dotc.Run.compileUnits(Run.scala:287)
        at dotty.tools.dotc.Run.compileSources(Run.scala:220)
        at dotty.tools.dotc.Run.compile(Run.scala:204)
        at dotty.tools.dotc.Driver.doCompile(Driver.scala:39)
        at dotty.tools.dotc.Driver.process(Driver.scala:199)
        at dotty.tools.dotc.Driver.process(Driver.scala:167)
        at dotty.tools.dotc.Driver.process(Driver.scala:179)
        at dotty.tools.dotc.Driver.main(Driver.scala:209)
        at dotty.tools.dotc.Main.main(Main.scala)
```
</details>

--- a/compiler/src/dotty/tools/dotc/core/Denotations.scala
+++ b/compiler/src/dotty/tools/dotc/core/Denotations.scala
@@ -403,7 +403,7 @@ object Denotations {
           }
         case denot1: SingleDenotation =>
           if (denot1 eq denot2) denot1
-          else if (denot1.matches(denot2)) mergeSingleDenot(denot1, denot2)
+          else if denot1.matches(denot2) then mergeSingleDenot(denot1, denot2)
           else NoDenotation
       }
 
@@ -438,8 +438,11 @@ object Denotations {
             else defn.RootClass)
 
         def isHidden(sym: Symbol) = sym.exists && !sym.isAccessibleFrom(pre)
-        val hidden1 = isHidden(sym1)
-        val hidden2 = isHidden(sym2)
+        // In typer phase filter out denotations with symbols that are not
+        // accessible. After typer, this is not possible since we cannot guarantee
+        // that the current owner is set correctly. See pos/14660.scala.
+        val hidden1 = isHidden(sym1) && ctx.isTyper
+        val hidden2 = isHidden(sym2) && ctx.isTyper
         if hidden1 && !hidden2 then denot2
         else if hidden2 && !hidden1 then denot1
         else

--- a/compiler/src/dotty/tools/dotc/typer/Checking.scala
+++ b/compiler/src/dotty/tools/dotc/typer/Checking.scala
@@ -537,7 +537,7 @@ object Checking {
     checkCombination(Private, Protected)
     checkCombination(Abstract, Override)
     checkCombination(Private, Override)
-    checkCombination(Private, Opaque)
+    if sym.isType && !sym.isClass then checkCombination(Private, Opaque)
     checkCombination(Lazy, Inline)
     // The issue with `erased inline` is that the erased semantics get lost
     // as the code is inlined and the reference is removed before the erased usage check.

--- a/compiler/src/dotty/tools/dotc/typer/Checking.scala
+++ b/compiler/src/dotty/tools/dotc/typer/Checking.scala
@@ -537,6 +537,7 @@ object Checking {
     checkCombination(Private, Protected)
     checkCombination(Abstract, Override)
     checkCombination(Private, Override)
+    checkCombination(Private, Opaque)
     checkCombination(Lazy, Inline)
     // The issue with `erased inline` is that the erased semantics get lost
     // as the code is inlined and the reference is removed before the erased usage check.

--- a/compiler/src/scala/quoted/runtime/impl/QuoteMatcher.scala
+++ b/compiler/src/scala/quoted/runtime/impl/QuoteMatcher.scala
@@ -441,7 +441,7 @@ object QuoteMatcher {
   }
 
   /** Result of matching a part of an expression */
-  private opaque type Matching = Option[Tuple]
+  private type Matching = Option[Tuple]
 
   private object Matching {
 

--- a/docs/_docs/reference/other-new-features/opaques-details.md
+++ b/docs/_docs/reference/other-new-features/opaques-details.md
@@ -51,6 +51,8 @@ object o:
 def id(x: o.T): o.T = x
 ```
 
+Opaque type aliases cannot be `private` and cannot overridden in subclasses.
+
 ## Type Parameters of Opaque Types
 
 Opaque type aliases can have a single type parameter list. The following aliases

--- a/tests/neg/i14660.scala
+++ b/tests/neg/i14660.scala
@@ -1,2 +1,8 @@
 class Bar:
   private opaque type Baz = Int // error
+
+  private object Foo:
+    opaque type O = Int // OK
+
+  val x: Baz = 1
+

--- a/tests/neg/i14660.scala
+++ b/tests/neg/i14660.scala
@@ -1,0 +1,2 @@
+class Bar:
+  private opaque type Baz = Int // error

--- a/tests/pos/i14660.scala
+++ b/tests/pos/i14660.scala
@@ -1,0 +1,6 @@
+trait Foo:
+  class Bar:
+    private[Foo] opaque type Baz = Int
+
+  def foo: Bar#Baz
+

